### PR TITLE
Sync: Don't clear $this->just_published completely

### PR DESCRIPTION
**Pull request by lezama**
Fixes: https://github.com/Automattic/io/issues/19#issuecomment-281814105

Some plugins could hook into `save_post` and `wp_insert_post` from within the hook.

In that case we were clearing `$this->just_published` without waiting for the original post from being inserted.

**How to test:**
Publishing a post containing an embedded recipe using the WP Recipe Maker plugin replicates the scenario pointed above.
See: https://github.com/Automattic/io/issues/19#issuecomment-281814105

### CODE REVIEW

**Comment by aduth** [comment]
It appears there were a few build variations that failed, though it's unclear why they would and I'm unable to retrieve the logs. Perhaps just needs a build restart?

**Comment by aduth** [comment]
I'm not as familiar with when `send_published` is called and how `just_published` is assigned, but the logic of avoiding clearing all "just published" when sending a single post (in lieu if just clearing the one sent) sounds reasonable to me.

I've also run these changes in the context of my Jetpack site configured to Publicize through my sandbox and confirmed in the before/after case of the changes applied that both posts with and without a recipe Publicized successfully. 👍

File: sync/class.jetpack-sync-module-posts.php [hunk 1]
```diff
@@ -227,47 +226,42 @@ function filter_post_content_and_add_links( $post_object ) {
...
+
+			/**
+			 * Action that gets synced when a post type gets published.
+			 *
+			 * @since 4.4.0
+			 *
+			 * @param int post_id
+			 * @param mixed array post flags that are added to the post
+			 */
+
```
Comment (jeherve) [inline sync/class.jetpack-sync-module-posts.php]: Looks like you have an extra empty line that will trip the parser.

File: sync/class.jetpack-sync-sender.php [hunk 2]
```diff
@@ -77,6 +77,7 @@ private function continue_full_sync_enqueue() {
 	}
 
 	public function do_sync() {
+		do_action( 'jetpack_sync_before_do_sync' );
```
Comment (jeherve) [inline sync/class.jetpack-sync-sender.php]: Could you add a docblock here? Thanks!

File: sync/class.jetpack-sync-module-posts.php [hunk 3]
```diff
@@ -244,25 +250,33 @@ public function send_published() {
...
+		 * Action that gets synced when a post type gets published.
+		 *
+		 * @since 4.4.0
+		 *
+		 * @param int, post_id
+		 * @param mixed array post flags that are added to the post
+		 */
+		do_action( 'jetpack_published_post', $post_ID, $flags );
+
+		if ( ( $post_ID = array_search( 'strawberry', $this->just_published ) ) !== false) {
```
Comment (lezama) [inline sync/class.jetpack-sync-module-posts.php]: yum, strawberries!

File: sync/class.jetpack-sync-module-posts.php [hunk 4]
```diff
@@ -22,14 +23,14 @@ public function set_defaults() {
...
-		add_action( 'wp_insert_post', array( $this, 'send_published'), 11, 3 );
+		$this->action_handler = $callable;
+		add_action( 'wp_insert_post', array( $this, 'wp_insert_post' ), 10, 3 );
 		add_action( 'deleted_post', $callable, 10 );
 		add_action( 'jetpack_publicize_post', $callable );
 		add_action( 'jetpack_published_post', $callable, 10, 2 );
 		add_action( 'transition_post_status', array( $this, 'save_published' ), 10, 3 );
 		add_filter( 'jetpack_sync_before_enqueue_wp_insert_post', array( $this, 'filter_blacklisted_post_types' ) );
-
+
```
Comment (lezama) [inline sync/class.jetpack-sync-module-posts.php]: we need to remove this spaces, waiting for tests to pass first 😓

File: tests/php/modules/publicize/test_class.publicize.php [hunk 5]
```diff
@@ -28,7 +28,7 @@ public function test_fires_jetpack_publicize_post_on_save_as_published() {
 		$this->post->post_status = 'publish';
 
 		wp_insert_post( $this->post->to_array() );
-
+
```
Comment (lezama) [inline tests/php/modules/publicize/test_class.publicize.php]: more spaces to clean

**Comment by lezama** [comment]
@aduth after 30 backs and forwards with @enejb :) I think it is good to go now. Could you give it another try?

File: sync/class.jetpack-sync-sender.php [hunk 6]
```diff
@@ -77,6 +77,12 @@ private function continue_full_sync_enqueue() {
 	}
 
 	public function do_sync() {
+		/**
```
Comment (enejb) [inline sync/class.jetpack-sync-sender.php]: This comment needs to go.

**Comment by enejb** [comment]
I lest a minor comment. I think once update the change this PR is ready to go.

**Pull request merged**